### PR TITLE
style(frontend): activity incomplete transaction list

### PR DESCRIPTION
**Pull request by BonomoAlessandro**
# Motivation

On the activity page we want to show a disclaimer to the user if he has active tokens that do not have an index canister yet. The disclaimer should list all affected tokens.

# Changes

- displays disclaimer

# Tests

**before:**
image

**after:**
image

### CODE REVIEW

File: src/frontend/src/lib/derived/network-tokens.derived.ts [hunk 1]
```diff
@@ -36,3 +39,17 @@ export const combinedDerivedSortedNetworkTokensUi: Readable<TokenUi[]> = derived
 			$exchanges
 		})
 );
+
+/**
+ * All user-enabled tokens matching the selected network or chain fusion that do not have an index canister.
+ */
+export const enabledNetworkTokensWithoutIndexCanister: Readable<Token[]> = derived(
```
Comment (peterpeterparker) [inline src/frontend/src/lib/derived/network-tokens.derived.ts:46]: - Is there plan to reuse this information anywhere else? If not, I would scope the information to the component. I think a lean approach about adding new stores is for the best given our performance issues.

- Why considering ETH and BTC? Should you just iterate on `enabledIcrcTokens` or `enabledIcrcTokensNoCk`?

File: src/frontend/src/lib/i18n/en.json [hunk 2]
```diff
@@ -803,6 +803,9 @@
 		},
 		"info": {
 			"btc_transactions": "BTC transaction information is obtained from central third parties and should be independently verified."
+		},
+		"warning": {
+			"incomplete_transaction_list": "Transaction list is incomplete because $token_list don't have an index canister to provide transaction information."
```
Comment (peterpeterparker) [inline src/frontend/src/lib/i18n/en.json:808]: `Index` with a capital letter?

File: src/frontend/src/lib/components/transactions/AllTransactions.svelte [hunk 3]
```diff
@@ -1,13 +1,30 @@
...
 	import PageTitle from '$lib/components/ui/PageTitle.svelte';
+	import { enabledNetworkTokensWithoutIndexCanister } from '$lib/derived/network-tokens.derived';
 	import { i18n } from '$lib/stores/i18n.store';
+	import { replacePlaceholders } from '$lib/utils/i18n.utils.js';
+
+	let enabledTokensWithoutCanister: Token[];
+	$: enabledTokensWithoutCanister = $enabledNetworkTokensWithoutIndexCanister;
+
+	let tokenList: string;
+	$: tokenList = enabledTokensWithoutCanister.map((token) => `$${token.symbol}`).join(', ');
```
[the remark this answers is not in the document]
Comment (BonomoAlessandro) [reply src/frontend/src/lib/components/transactions/AllTransactions.svelte:14]: In the Figma it is explicitly with the prefix `$`.

File: src/frontend/src/lib/components/transactions/AllTransactions.svelte [hunk 4]
```diff
@@ -1,13 +1,30 @@
...
 </script>
 
 <div class="flex flex-col gap-5">
 	<PageTitle>{$i18n.activity.text.title}</PageTitle>
 
+	{#if tokenList}
```
[the remark this answers is not in the document]
Comment (BonomoAlessandro) [reply src/frontend/src/lib/components/transactions/AllTransactions.svelte:20]: This check will return false if the `tokenList` is `undefined`, `null` or `empty`. If no token were found than the `tokenList` will be empty and the MessageBox will not be displayed. It works as expected.

File: src/frontend/src/lib/derived/network-tokens.derived.ts (hunk 1 above)
Comment (BonomoAlessandro) [reply src/frontend/src/lib/derived/network-tokens.derived.ts:46]: @peterpeterparker 
No, at the moment it is not planed to reuse it anywhere else. I guess you are right, i should move the implementation into the component. 

Why not? Isn't it possible that an ETH or BTC does not have an Index canister?

File: src/frontend/src/lib/components/transactions/AllTransactions.svelte (hunk 4 above)
[the remark this answers is not in the document]
Comment (peterpeterparker) [reply src/frontend/src/lib/components/transactions/AllTransactions.svelte:32]: > or empty.

I don't think so.

Capture d’écran 2024-11-22 à 09 07 51

File: src/frontend/src/lib/derived/network-tokens.derived.ts (hunk 1 above)
Comment (peterpeterparker) [reply src/frontend/src/lib/derived/network-tokens.derived.ts]: > No, at the moment it is not planed to reuse it anywhere else. I guess you are right, i should move the implementation into the component.

I would say let's do that if that works for you?

> Isn't it possible that an ETH or BTC

Canisters only existing on the Internet Computer that's why I'm wondering why we are considering Ethereum and Bitcoin, unless e.g. ckBTC or ckETH are listed explicitely within this derived stores but, I don't think it's the case?

File: src/frontend/src/lib/components/transactions/AllTransactions.svelte (hunk 4 above)
[the remark this answers is not in the document]
Comment (BonomoAlessandro) [reply src/frontend/src/lib/components/transactions/AllTransactions.svelte:32]: @peterpeterparker 
But your test is wrong. The `tokenList` is not an array but the list of the tokens as `string`. It is an empty string.

File: src/frontend/src/lib/derived/network-tokens.derived.ts (hunk 1 above)
Comment (BonomoAlessandro) [reply src/frontend/src/lib/derived/network-tokens.derived.ts]: @peterpeterparker 
That's ok for me, i just moved the implementation into the component.

Hmm, i guess you are right.

File: src/frontend/src/lib/components/transactions/AllTransactions.svelte (hunk 4 above)
[the remark this answers is not in the document]
Comment (peterpeterparker) [reply src/frontend/src/lib/components/transactions/AllTransactions.svelte:32]: Ahhh my bad! Sorry!

Nitpick then: can you use `notEmptyString(tokensList)`

File: src/frontend/src/lib/components/transactions/AllTransactions.svelte [hunk 5]
```diff
@@ -1,10 +1,23 @@
...
 	import MessageBox from '$lib/components/ui/MessageBox.svelte';
 	import PageTitle from '$lib/components/ui/PageTitle.svelte';
+	import { enabledNetworkTokens } from '$lib/derived/network-tokens.derived';
 	import { testnetsEnabled } from '$lib/derived/settings.derived';
 	import { i18n } from '$lib/stores/i18n.store';
+	import type { TokenUi } from '$lib/types/token';
+	import { replacePlaceholders } from '$lib/utils/i18n.utils.js';
+
+	let enabledTokensWithoutCanister: Token[];
+	$: enabledTokensWithoutCanister = $enabledNetworkTokens.filter(
```
Comment (peterpeterparker) [inline src/frontend/src/lib/components/transactions/AllTransactions.svelte:15]: @AntonioVentilii implemented a workaround last week that causes tokens with an index canister in error to fall back to using the ledger only. As a result, `transactions === null` can now have two different root causes, which are displayed differently in the UI. Can you both discuss how you'd like to handle this?

Comment (AntonioVentilii) [reply src/frontend/src/lib/components/transactions/AllTransactions.svelte:15]: Discussed offline with @BonomoAlessandro , and IIRC, he said that there will be two different info boxes for the tokens without index canisters and the ones with index canister not working. So it makes sense to generalize/improve the workaround I did

Did I understand correctly, @BonomoAlessandro ?

Comment (BonomoAlessandro) [reply src/frontend/src/lib/components/transactions/AllTransactions.svelte:15]: @AntonioVentilii yess exactly.

Comment (BonomoAlessandro) [reply src/frontend/src/lib/components/transactions/AllTransactions.svelte]: @peterpeterparker @AntonioVentilii 
I just added the warning message box that displays the enabled token that do not have an index canister yet. I also provided a test and updated the PR description.
The second warning message box which displays the enabled token with a broken index canister will come with antoher PR.

File: src/frontend/src/icp/validation/ic-token.validation.ts [hunk 6]
```diff
@@ -29,3 +30,9 @@ export const isIcCkToken = (token: Token): token is IcCkToken => {
 
 export const isNotIcCkToken = (token: Token): token is Exclude<Token, IcCkToken> =>
 	!isIcCkToken(token);
+
+export const hasIndexCanister = (token: IcToken): token is IcToken & IcCanistersStrict =>
```
[the remark this answers is not in the document]
Comment (BonomoAlessandro) [reply src/frontend/src/icp/validation/ic-token.validation.ts:34]: @peterpeterparker I just implemented some tests for the new functions.

File: src/frontend/src/icp/validation/ic-token.validation.ts [hunk 7]
```diff
@@ -29,3 +30,8 @@ export const isIcCkToken = (token: Token): token is IcCkToken => {
...
+	nonNullish(token) && isIcTokenCanistersStrict(token);
+
+export const hasNoIndexCanister = (token: IcToken): token is IcToken => !hasIndexCanister(token);
```
Comment (peterpeterparker) [inline src/frontend/src/icp/validation/ic-token.validation.ts:37]: If a token has an index canister, it's also an `IcToken`. Maybe better to just type the return as a `boolean` and do not specify assertion here?

**Review by peterpeterparker — COMMENTED** [review]
Looking good. Just a minor last comment.

File: src/frontend/src/icp/validation/ic-token.validation.ts (hunk 7 above)
Comment (BonomoAlessandro) [reply src/frontend/src/icp/validation/ic-token.validation.ts:37]: @peterpeterparker hmm, yes might be the easiest.

**Review by peterpeterparker — APPROVED** [review]
LGTM, thx

**Pull request merged**